### DISCUSSION TITLE
.NET MAUI updates in .NET 9 Preview 7

### DISCUSSION THREAD

**Discussion by davidortinau** [dotnet/maui#24219] [Announcements]
Here's a summary of what's new in .NET MAUI in this preview release:

* [Introduction of `HybridWebview`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#introduction-of-hybridwebview)
* [New `TitleBar` Control and `Window.TitleBar` for Windows](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#new-titlebar-control-and-windowtitlebar-for-windows)
* [`CollectionView` & `CarouselView` improvements with a new opt-in handler for iOS and Mac Catalyst](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#collectionview--carouselview-improvements-with-a-new-opt-in-handler-for-ios-and-mac-catalyst)
* [Ability to bring a `Window` to the foreground with `ActivateWindow`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#activatewindow-added-to-bring-a-window-to-foreground)
* [`BackButtonBehavior` `OneWay` binding mode](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#backbuttonbehavior-oneway-binding-mode)
* [`BlazorWebView` backward compatibility host address](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#blazorwebview-backward-compatibility-host-address)
* [Native Embedding improvements](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#native-embedding-improvements)
* [`MainPage` is Obsolete](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#mainpage-is-obsolete)
* [New Handler Disconnect Policy](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#new-handler-disconnect-policy)
* [New `ProcessTerminated` event on `WebView` Control](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#new-processterminated-event-on-webview-control)
* [New lifecycle methods for remote notifications on iOS & Mac Catalyst](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#new-lifecycle-methods-for-remote-notifications-on-ios--mac-catalyst)
* [Xcode Sync for CLI and Visual Studio Code](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#xcode-sync-for-cli-and-visual-studio-code)

.NET MAUI updates in .NET 9 Preview 7:
* [Release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md) 
* [What's new in .NET MAUI in .NET 9](https://learn.microsoft.com/dotnet/maui/whats-new/dotnet-9) documentation.
* [GitHub Release](https://aka.ms/maui9p7)

.NET 9 Preview 7:
* [Discussion](https://aka.ms/dotnet/9/preview7)
* [Release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/README.md)

**Comment by OwnageIsMagic** [comment]
[New lifecycle methods for remote notifications on iOS & Mac Catalyst](https://github.com/dotnet/maui/discussions/24219#new-lifecycle-methods-for-remote-notifications-on-ios--mac-catalyst) link broken

**Reply by davidortinau** [reply]
Thx, fixed

**Comment by ljzj2** [comment]
what is a native embedding ? when to use it?

**Reply by Goksly** [reply]
If you click on the link, it takes you to the change which contains the link to what Native Embedded is. Go forage my child!

**Comment by maonaoda** [comment]
Is there any way to create a custom nuget for Maui locally?
For example, I want to make some modifications to Microsoft.Maui.Controls.Core, but the nugets I generated locally are all 9.0.0-dev.24419.1, which cannot be used with the latest 9.0.0-preview.7.24407.4

**Reply by MartyIX** [reply]
Would this work: *You can build MAUI locally, put nugets to a folder, add that folder to `NuGet.config` as a source.* ?

**Reply by maonaoda** [reply]
@MartyIX 
Thanks for your reply. I can compile and generate nuget locally, but this will generate all packages. I just want to generate a package similar to 9.0.0-preview.7.24407.**5** based on 9.0.0-preview.7.24407.4.

However, this is no longer a problem, I have uploaded the generated package to a private nuget feed

**Reply by MartyIX** [reply]
I think that one needs to understand a bit Arcade to generate such packages. There is a msbuild property to generate preview packages IMO. @rmarinho knows how this stuff work.

**Reply by maonaoda** [reply]
@MartyIX 
Yes, I tried to change it, but I was not familiar with many things and failed.

The original plan was to use Net8, but due to some bug fixes, Net9 had to be used, which has been putting the migration of the app on hold.
In fact, if some pr can be merged quickly, I don't need to compile package like this.

**Comment by david-nyaoso** [comment]
With the latest dotnet 9 preview 7 and VS 2022 17.12.0 Preview 1.0,  the deploy checkbox in configuration manager dialog is grayed out in a Maui-Blazor-Web app when (targeting) building for Windows. When debugging, the app starts and runs 'normally'. But code such as global::Windows.ApplicationModel.Package.Current.InstalledLocation throws an exception because the was not 'deployed' (no windows package). Running the same code targeting dotnet 8 just works!

**Comment by DDHSchmidt** [comment]
> [CollectionView & CarouselView improvements with a new opt-in handler for iOS and Mac Catalyst](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#collectionview--carouselview-improvements-with-a-new-opt-in-handler-for-ios-and-mac-catalyst)

`The type or namespace name 'Items2' does not exist in the namespace 'Microsoft.Maui.Controls.Handlers' (are you missing an assembly reference?)`
![image](https://github.com/user-attachments/assets/ccf4369a-8d43-4a68-9010-3887e6121b64)



Output of dotnet workload list:
```
android                       SDK 9.0.100-preview.7, VS 17.12.35209.166
aspire                        SDK 9.0.100-preview.7
ios                           SDK 9.0.100-preview.7, VS 17.12.35209.166
maccatalyst                   SDK 9.0.100-preview.7, VS 17.12.35209.166
maui                          SDK 9.0.100-preview.7
maui-windows                  SDK 9.0.100-preview.7, VS 17.12.35209.166
```
dotnet --version:
`9.0.100-preview.7.24407.12`

What's the magic sauce I'm missing to enjoy the new Handler-goodness, @davidortinau  ?

**Reply by MichelePannacciTerex** [reply]
> > [CollectionView & CarouselView improvements with a new opt-in handler for iOS and Mac Catalyst](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#collectionview--carouselview-improvements-with-a-new-opt-in-handler-for-ios-and-mac-catalyst)
> 
> `The type or namespace name 'Items2' does not exist in the namespace 'Microsoft.Maui.Controls.Handlers' (are you missing an assembly reference?)` ![image](https://private-user-images.githubusercontent.com/67259411/363639964-ccf4369a-8d43-4a68-9010-3887e6121b64.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgyMTYyNDgsIm5iZiI6MTcyODIxNTk0OCwicGF0aCI6Ii82NzI1OTQxMS8zNjM2Mzk5NjQtY2NmNDM2OWEtOGQ0My00YTY4LTkwMTAtMzg4N2U2MTIxYjY0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA2VDExNTkwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI4YzAyMmIyYjc4YWI1ZGUzN2ZkNjQ4NjBiY2FhMTUzZTM2ZjQ2ODUwZThlODlkMTJkZjczNDQyYzIwOWVjOGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5OzUH89b5bgCvwHp698rQlUBGLsar8fW2U7ojSatwf0)
> 
> Output of dotnet workload list:
> 
> ```
> android                       SDK 9.0.100-preview.7, VS 17.12.35209.166
> aspire                        SDK 9.0.100-preview.7
> ios                           SDK 9.0.100-preview.7, VS 17.12.35209.166
> maccatalyst                   SDK 9.0.100-preview.7, VS 17.12.35209.166
> maui                          SDK 9.0.100-preview.7
> maui-windows                  SDK 9.0.100-preview.7, VS 17.12.35209.166
> ```
> 
> dotnet --version: `9.0.100-preview.7.24407.12`
> 
> What's the magic sauce I'm missing to enjoy the new Handler-goodness, @davidortinau ?

Did you find out?

**Reply by DDHSchmidt** [reply]
> Did you find out?

[Yeah, my tower of MAUI-workarounds got the better of me](https://github.com/dotnet/maui/discussions/24698#discussioncomment-10627814)
